### ISSUE TITLE
@liouxy Yes. foreground:background is 1:1 . Since the mask is no longer rectangular, it is difficult to calculate the discounted weights proposed in CVPR paper. So the final solution is just a simple pixel-wise loss over input and output.

### ISSUE THREAD

**Issue by huangqianfirst** [JiahuiYu/generative_inpainting#172]
_Originally posted by @JiahuiYu in https://github.com/JiahuiYu/generative_inpainting/issues/62#issuecomment-402377802_

**Comment by JiahuiYu** [comment]
Do you have any question?

**Comment by huangqianfirst** [comment]
Thansks for your great work.
I have some questions:
Q1:For the config.COARSE_L1_ALPHA, v1 wrote like below.So is the l1_alpha need to change? keep old or set it to zero.
l1_alpha = config.COARSE_L1_ALPHA
        losses['l1_loss'] = l1_alpha * tf.reduce_mean(tf.abs(batch_pos- x1)
        if not config.PRETRAIN_COARSE_NETWORK:
            losses['l1_loss'] += tf.reduce_mean(tf.abs(batch_pos - x2)
Q2:For the GAN input.Is the input of discriminator,
batch_pos_neg = tf.concat([batch_pos, batch_complete], axis=0)
Or 
batch_pos_neg = tf.concat([batch_pos, batch_predicted], axis=0)
? what's the difference between use batch_complete and use batch_predicted?
And it's right batch_pos_neg should concat with mask?

**Comment by JiahuiYu** [comment]
No problem. Feel free to leave any further question here.

**Comment by huangqianfirst** [comment]
hi, @JiahuiYu , I still have some questiones.
I've trained a model a few days on the whole dataset(Places2) with iters = 92000,loss is 0.056. And the result on the test dataset is not good.It may because the input of D is batch predicted, not batch_complete or iters is not long enough.
![00000001_out_1](https://user-images.githubusercontent.com/11473038/48245625-b4eb2680-e426-11e8-894c-d0570677f1f4.png)
And the ,I want to finetune in more small dateset(For a quick check if my code it's right). 5000(contain 50 test iamges) airfiled images in Places2. LR has been changed to 1e-5.I think it may lead loss to down more. Loss it to 0.03, reduce lr can't make loss less. As you can see result is still not good. Overfitting is not exist i think.sencod is test on train data.
![00000001_out_4](https://user-images.githubusercontent.com/11473038/48245751-6db16580-e427-11e8-9130-0f842db2b209.png)
![default](https://user-images.githubusercontent.com/11473038/48246706-6b510a80-e42b-11e8-9c45-68479302c139.PNG)

Q1. what's the better training method?Training first on the wholes datasets, and then finetune? How many iters and what can i gain form loss changes to confirm when to stop wehther result is good or not  . 
Q2. How can i get a quick check whether my model works or not.
Q3. It's finetune must on a wonderful model result?Or on a convergent model but not best.

**Comment by huangqianfirst** [comment]
hi, @JiahuiYu , something confused me.
Q1.About the g_loss.
At first,
d_loss : 0.99485266. This value will decrease fast.
g_loss( without l1_loss) :  eg.-0.06371535，This value will grow fast, to someone >0.
In D_net： g_loss = -tf.reduce_mean(neg), it suppose to grow.
But we know l1_loss will decrease to zero.
so  losses['g_loss'] = g_loss +losses['l1_loss'], what's the trend of this value ?a contradiction ?
Q2.About code.
And could please check for my code .

batch_pos_neg has been concat with mask.
pos_neg_global = self.build_SNgan_discriminator(batch_pos_neg, training=training, reuse=reuse)
## SN_discriminator is leak_relu(cond_sn) every conv.
pos_global, neg_global = tf.split(pos_neg_global, 2, axis = 0)
g_loss, d_loss = self.gan_hinge_loss(pos_global, neg_global)
## 
losses['g_loss'] = g_loss
losses['d_loss'] = d_loss

the complete is losses['g_loss'] = g_loss +losses['l1_loss']
Has any problem? Hyperparamter has not to set ?

**Comment by JiahuiYu** [comment]
It is normal.

**Comment by huangqianfirst** [comment]
After 12 hours training,  losses['g_loss'] has drop form 0.9 to -275, something smaller. Does it still normal.
PS.Could Please check the code about loss for me? Thanks.

**Comment by JiahuiYu** [comment]
@huangqianfirst Have you succeed in training with DeepFill v1, as released in this code base? You posted code seems correct. I think at least you should get some positive results with your own dataset (I guess it is not public dataset?).

**Comment by JiahuiYu** [comment]
It is not normal. However, I don't know the reason. It seems your cases is relatively easy, the inpainting network should perform much better.

**Comment by huangqianfirst** [comment]
I'm sorry to bother you.But I still has some questions.
The change of loss.
the loss of D_net is:
g_loss, d_loss = self.gan_hinge_loss(pos_global, neg_global)
the complete loss of G loss is : 
losses['g_loss'] = a1 *  g_loss +a2 * losses['l1_loss']
Q1.What happens to this g_loss of D_net , it may around 0 ,or would drop to -12 or more. what's the minimum? What about d_loss ?
Q2.the a1, a2 that weights of complete losses['g_loss'] is 1 ?
I'm looking forward to your answer.It will help me check if my code it's wrong or not.

**Comment by huangqianfirst** [comment]
I'm sorry to bother you.But I still has some questions.
The change of loss.
the loss of D_net is:
g_loss, d_loss = self.gan_hinge_loss(pos_global, neg_global)
the complete loss of G loss is : 
losses['g_loss'] = a1 *  g_loss +a2 * losses['l1_loss']
Q1.What happens to this g_loss of D_net , it can drop to -9. Of cource ,this result is awful. what's the range of g_loss, d_loss?
Q2.the a1, a2 that weights of complete losses['g_loss'] is 1 ?
I'm looking forward to your answer.It will help me check if my code it's wrong or not.
